### ISSUE TITLE
Can this library be used for similarity metric learning

### ISSUE THREAD

**Issue by lesshaste** [metric-learn/metric-learn#98]
I have a set of vectors and for each pair of these vectors I have a distance (which is not Euclidean).  I would like to embed the vectors into Euclidean space so that they are more likely to be close in R^d if they are close under the original measure of distance. I believe this is called similarity metric learning.

Is there a way to use this metric-learn library for this setup? It seems from the examples that you need a label for each vector, rather than  a distance for each pair as input.  Is there some way to get round this?

**Comment by thomasniebler** [comment]
Hi,
there are some algorithms (like e.g. LSML or a variant of a learning-to-rank SVM) that use relative distance comparisons to learn a distance metric. However, I doubt that you could directly use any of the currently implemented algorithms to solve this task without any further coding.
Still I think that this shouldn't be much of a problem. Use the distances from your non-Euclidean vector space as input for your choice of metric learning algorithm, but as vector representations, use a randomly initialized matrix (similar to Word2Vec or GloVe), which you update after each learning step with your learned transformation matrix. After this, you should probably end up with embeddings according to your distances. Be aware that this is only a rough sketch of how it could maybe work.
Best
Thomas

**Comment by perimosocordiae** [comment]
In your case, it sounds like you've already learned a similarity metric. If you have the full NxN matrix of distances, you can use an embedding method to learn a transformation of your original data to the desired space.

If you want a linear transform from original to embedded space, there are [plenty of options](https://stat.columbia.edu/~cunningham/pdf/CunninghamJMLR2015.pdf) but I would typically use Locality Preserving Projections (LPP). Here's an implementation I wrote a while back:
https://github.com/all-umass/graphs/blob/4fbeb025dfe33340335f34300f58dd3809228822/graphs/mixins/embed.py#L31

If you don't care about getting a linear transformation and just want the final embedded vectors, you could use kernel PCA (with `kernel='precomputed'`): http://scikit-learn.org/stable/modules/generated/sklearn.decomposition.KernelPCA.html

**Comment by bellet** [comment]
What you are looking for sounds more like "metric embedding", such as the classic MDS algorithm:
http://scikit-learn.org/stable/modules/generated/sklearn.manifold.MDS.html

**Comment by lesshaste** [comment]
Thank you for the great replies. The problem with the scikit learn methods is that most of them don't seem to support the concept of a training and test set. That is most of the methods  don't have "transform" method.  In fact I think only manifold.Isomap and manifold.LocallyLinearEmbedding do.

**Comment by bellet** [comment]
You might be interested in this:
https://papers.nips.cc/paper/2461-out-of-sample-extensions-for-lle-isomap-mds-eigenmaps-and-spectral-clustering.pdf

Otherwise as suggested above you would need to use a parametric approach, e.g. based on learning a linear transformation.
